### ISSUE TITLE
"layer.get_weights()" does not work properly after the first use of "layer.set_weights()"

### ISSUE THREAD

**Issue by MohammadSamragh** [awslabs/keras-apache-mxnet#138]
In the following code, I am reading the weights of the first layer before and after training. I compute the  difference between the weights before and after training. The difference is non-zero as expected:
```

model=get_model()
w_before,b=model.layers[0].get_weights()
#w_before=np.random.rand(w_before.shape[0],w_before.shape[1],w_before.shape[2],w_before.shape[3])
#model.layers[0].set_weights([w_before,b])
opt = keras.optimizers.SGD(lr=0.001, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy',optimizer=opt,metrics=['accuracy'])
model.fit(X_train,Y_train,...)
w_after=model.layers[0].get_weights()[0]
difference=np.sum(np.absolute(w_before-w_after))
print difference
```

However, when I use the "set_weights()" function (by uncommenting line 3 and 4 of the above code), the "get_weights()" function does not work properly anymore and the difference becomes zero.

I read the "set_value()" function in the "mxnet_backend.py" file. It uses "mxnet.symbol.bind()" to set the weights. Is it possible that "bind" is causing this? Is there any alternative way of setting values to variables?

**Comment by roywei** [comment]
@MohammadSamragh thanks for this feedback, the problem may also occur at get_weights after training, not necessary failed at set_weight and bind. While we look into this, may I know whats the usecase here? and besides the difference=0 problem, does setting the weights affect final loss & accuracy compare with TensorFlow backend?
It will be helpful to debug if you can provide a reproducible code including maybe a simplified version of your model training script.
Thanks!

**Comment by MohammadSamragh** [comment]
Thank for the response. I did try get_weights after training (without set_weights before training) and it worked properly (difference was not zero).

The use case is when you want to load a pre-trained model, fine-tune it, then save the fine-tuned network. I assume set_weights is inherently used within model.load_weights. After loading a pre-trained model, if I fine-tune the model, the accuracy and the loss improve as the training proceeds. However, when I want to save the fine-tuned model, it saves the original weights that were initially loaded, as if there was no fine-tuning.

Here is a re-producible code:
```
import numpy as np
import keras
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential, Model
from keras.layers import Dense, Activation, Input, Lambda, Conv2D, MaxPooling2D, Flatten, Dropout
from keras.layers.normalization import BatchNormalization
from keras import backend as K
from keras.engine.topology import Layer
from keras.datasets import cifar10
from keras.models import load_model
from keras.utils import np_utils
from keras.regularizers import l2


epochs=1
learning_rate=0.001
batch_size=100

from keras.callbacks import Callback

def get_model():
#cite this: https://github.com/geifmany/cifar-vgg
	model = Sequential()
	weight_decay = 0.0005
	if K.image_data_format=='channels_last':
		input_shape=[32,32,3]
	else:
		input_shape=[3,32,32]
	model.add(Conv2D(64, (3, 3), padding='same',
	                 input_shape=input_shape,kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())
	model.add(Dropout(0.3))

	model.add(Conv2D(64, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())

	model.add(MaxPooling2D(pool_size=(2, 2)))

	model.add(Conv2D(128, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())
	model.add(Dropout(0.4))

	model.add(Conv2D(128, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())

	model.add(MaxPooling2D(pool_size=(2, 2)))

	model.add(Conv2D(256, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())
	model.add(Dropout(0.4))

	model.add(Conv2D(256, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())
	model.add(Dropout(0.4))

	model.add(Conv2D(256, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())

	model.add(MaxPooling2D(pool_size=(2, 2)))


	model.add(Conv2D(512, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())
	model.add(Dropout(0.4))

	model.add(Conv2D(512, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())
	model.add(Dropout(0.4))

	model.add(Conv2D(512, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())

	model.add(MaxPooling2D(pool_size=(2, 2)))


	model.add(Conv2D(512, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())
	model.add(Dropout(0.4))

	model.add(Conv2D(512, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())
	model.add(Dropout(0.4))

	model.add(Conv2D(512, (3, 3), padding='same',kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())

	model.add(MaxPooling2D(pool_size=(2, 2)))
	model.add(Dropout(0.5))

	model.add(Flatten())
	model.add(Dense(512,kernel_regularizer=l2(weight_decay)))
	model.add(Activation('relu'))
	model.add(BatchNormalization())

	model.add(Dropout(0.5))
	model.add(Dense(10))
	model.add(Activation('softmax'))

	#model.load_weights(os.path.join(dataset,network,'pre_trained.h5'))

	return model









(X_train, y_train), (X_test, y_test) = cifar10.load_data()
print X_train.shape
X_train=X_train.astype(np.float32)
X_test=X_test.astype(np.float32)
Y_train = np_utils.to_categorical(y_train, 10)
Y_test = np_utils.to_categorical(y_test, 10)
mean = np.mean(X_train,axis=(0,1,2,3))
std = np.std(X_train, axis=(0, 1, 2, 3))
X_train = (X_train-mean)/(std+1e-7)
X_test = (X_test-mean)/(std+1e-7)

train_iter = ImageDataGenerator(
    featurewise_center=False,  # set input mean to 0 over the dataset
    samplewise_center=False,  # set each sample mean to 0
    featurewise_std_normalization=False,  # divide inputs by std of the dataset
    samplewise_std_normalization=False,  # divide each input by its std
    zca_whitening=False,  # apply ZCA whitening
    rotation_range=15,  # randomly rotate images in the range (degrees, 0 to 180)
    width_shift_range=0.1,  # randomly shift images horizontally (fraction of total width)
    height_shift_range=0.1,  # randomly shift images vertically (fraction of total height)
    horizontal_flip=True,  # randomly flip images
    vertical_flip=False)  # randomly flip images
# (std, mean, and principal components if ZCA whitening is applied).
train_iter.fit(X_train)
val_iter = ImageDataGenerator()
train_gen = train_iter.flow(X_train, Y_train,batch_size=batch_size)
val_gen = val_iter.flow(X_test, Y_test,batch_size=batch_size,shuffle=False)


#Let's train a model and save it:
model1=get_model()
opt = keras.optimizers.SGD(lr=0.001, decay=1e-6, momentum=0.9, nesterov=True)
model1.compile(loss='categorical_crossentropy',optimizer=opt,metrics=['accuracy'])
model1.fit_generator(train_gen,steps_per_epoch=50000/batch_size,nb_epoch=epochs,validation_data=val_gen,validation_steps=10000/batch_size,verbose=1)
model1.save_weights('pretrained.h5')


#Now let's read the pretrained model into a separate model and fine-tune it
model2=get_model()
#the following line inherently uses set_weights
model2.load_weights('pretrained.h5')
opt = keras.optimizers.SGD(lr=0.001, decay=1e-6, momentum=0.9, nesterov=True)
model2.compile(loss='categorical_crossentropy',optimizer=opt,metrics=['accuracy'])
_,acc=model2.evaluate(X_test,Y_test)
print 'evaluating the loaded model to make sure it loaded correctly... accuracy of the loaded model2 was', acc
model2.fit_generator(train_gen,steps_per_epoch=50000/batch_size,nb_epoch=epochs,validation_data=val_gen,validation_steps=10000/batch_size,verbose=1)

#now let's compare the weights of model with those of model2. They should be different, but they are not!:
print 'checking if the fine-tuned weights are different from the initial pre-trained weights...'
for i,l in enumerate(model1.layers):
	l2=model2.layers[i]
	if isinstance(l,Conv2D) or isinstance(l,Dense):
		w1,b1=l.get_weights()
		w2,b2=l2.get_weights()
		difference=np.sum(np.absolute(w1-w1))
		print 'difference for layer',l.name,'was',difference

#so if I save the fine-tuned model2 in a file named 'fine_tuned.h5', it will contain the exact same weights as in 'pretrained.h5'
model2.save_weights('fine_tuned.h5')

```
And here is the output of the code:

```
Using MXNet backend
/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
(50000, 3, 32, 32)
tucker.py:157: UserWarning: The semantics of the Keras 2 argument `steps_per_epoch` is not the same as the Keras 1 argument `samples_per_epoch`. `steps_per_epoch` is the number of batches to draw from the generator at each epoch. Basically steps_per_epoch = samples_per_epoch/batch_size. Similarly `nb_val_samples`->`validation_steps` and `val_samples`->`steps` arguments have changed. Update your method calls accordingly.
  model1.fit_generator(train_gen,steps_per_epoch=50000/batch_size,nb_epoch=epochs,validation_data=val_gen,validation_steps=10000/batch_size,verbose=1)
tucker.py:157: UserWarning: Update your `fit_generator` call to the Keras 2 API: `fit_generator(<keras.pre..., verbose=1, validation_data=<keras.pre..., steps_per_epoch=500, epochs=1, validation_steps=100)`
  model1.fit_generator(train_gen,steps_per_epoch=50000/batch_size,nb_epoch=epochs,validation_data=val_gen,validation_steps=10000/batch_size,verbose=1)
Epoch 1/1
/usr/local/lib/python2.7/dist-packages/mxnet/module/bucketing_module.py:408: UserWarning: Optimizer created manually outside Module but rescale_grad is not normalized to 1.0/batch_size/num_workers (1.0 vs. 0.01). Is this intended?
  force_init=force_init)
[23:36:12] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:107: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
500/500 [==============================] - 35s 71ms/step - loss: 2.7990 - acc: 0.2003 - val_loss: 2.5117 - val_acc: 0.2097
/usr/local/lib/python2.7/dist-packages/mxnet/module/bucketing_module.py:408: UserWarning: Optimizer created manually outside Module but rescale_grad is not normalized to 1.0/batch_size/num_workers (1.0 vs. 0.03125). Is this intended?
  force_init=force_init)
10000/10000 [==============================] - 4s 422us/step
evaluating the loaded model to make sure it loaded correctly... accuracy of the loaded model2 was 0.2097
tucker.py:169: UserWarning: The semantics of the Keras 2 argument `steps_per_epoch` is not the same as the Keras 1 argument `samples_per_epoch`. `steps_per_epoch` is the number of batches to draw from the generator at each epoch. Basically steps_per_epoch = samples_per_epoch/batch_size. Similarly `nb_val_samples`->`validation_steps` and `val_samples`->`steps` arguments have changed. Update your method calls accordingly.
  model2.fit_generator(train_gen,steps_per_epoch=50000/batch_size,nb_epoch=epochs,validation_data=val_gen,validation_steps=10000/batch_size,verbose=1)
tucker.py:169: UserWarning: Update your `fit_generator` call to the Keras 2 API: `fit_generator(<keras.pre..., verbose=1, validation_data=<keras.pre..., steps_per_epoch=500, epochs=1, validation_steps=100)`
  model2.fit_generator(train_gen,steps_per_epoch=50000/batch_size,nb_epoch=epochs,validation_data=val_gen,validation_steps=10000/batch_size,verbose=1)
Epoch 1/1
[23:36:49] src/executor/graph_executor.cc:711: Bucketing: data /activation_30_target1 has a shape [100,10], which is larger than already allocated shape [32,10]. Need to re-allocate. Consider putting default bucket key to be the bucket taking the largest input for better memory sharing.
[23:36:49] src/executor/graph_executor.cc:711: Bucketing: data /conv2d_14_input1 has a shape [100,3,32,32], which is larger than already allocated shape [32,3,32,32]. Need to re-allocate. Consider putting default bucket key to be the bucket taking the largest input for better memory sharing.
[23:36:49] src/executor/graph_executor.cc:711: Bucketing: data /activation_30_sample_weights1 has a shape [100], which is larger than already allocated shape [32]. Need to re-allocate. Consider putting default bucket key to be the bucket taking the largest input for better memory sharing.
500/500 [==============================] - 31s 61ms/step - loss: 2.3087 - acc: 0.3073 - val_loss: 2.5838 - val_acc: 0.2203
checking if the fine-tuned weights are different from the initial pre-trained weights...
difference for layer conv2d_1 was 0.0
difference for layer conv2d_2 was 0.0
difference for layer conv2d_3 was 0.0
difference for layer conv2d_4 was 0.0
difference for layer conv2d_5 was 0.0
difference for layer conv2d_6 was 0.0
difference for layer conv2d_7 was 0.0
difference for layer conv2d_8 was 0.0
difference for layer conv2d_9 was 0.0
difference for layer conv2d_10 was 0.0
difference for layer conv2d_11 was 0.0
difference for layer conv2d_12 was 0.0
difference for layer conv2d_13 was 0.0
difference for layer dense_1 was 0.0
difference for layer dense_2 was 0.0

```

**Comment by roywei** [comment]
fixed by #185
